src/test/func/sandbox/sandbox.cc

```diff
@@ -0,0 +1,234 @@
+     * pagemap and would not be used outside of the sandbox.
+     */
+    using InternalAlloc =
+      Allocator<never_init, no_op_init, MemoryProviderProxy>;
+
+    /**
+     * The start of the sandbox memory region.
+     */
+    void* start;
+
+    /**
+     * The end of the sandbox memory region
+     */
+    void* top;
+
+    /**
+     * State allocated in the sandbox that is shared between the inside and
+     * outside.
+     */
+    struct SharedState
+    {
+      /**
+       * The message queue for the allocator that lives outside of the
+       * sandbox but allocates memory inside.
+       */
+      struct RemoteAllocator queue;
+    } * shared_state;
+
+    /**
+     * The memory provider for this sandbox.
+     */
+    MemoryProviderStateMixin<NoOpPal> state;
+
+    /**
+     * The allocator for callers outside the sandbox to allocate memory inside.
+     */
+    ExternalAlloc alloc;
+
+    /**
+     * An allocator for callers inside the sandbox to allocate memory.
+     */
+    InternalAlloc* internal_alloc;
+
+    /**
+     * Constructor.  Takes the size of the sandbox as the argument.
+     */
+    Sandbox(size_t sb_size)
+    : start(alloc_sandbox_heap(sb_size)),
+      top(pointer_offset(start, sb_size)),
+      shared_state(new (start) SharedState()),
+      state(
+        pointer_offset(start, sizeof(SharedState)),
+        sb_size - sizeof(SharedState)),
+      alloc(state, SNMALLOC_DEFAULT_CHUNKMAP(), &shared_state->queue)
+    {
+      auto* state_proxy = static_cast<MemoryProviderProxy*>(
+        alloc.alloc(sizeof(MemoryProviderProxy)));
+      state_proxy->real_state = &state;
+      internal_alloc =
+        new (alloc.alloc(sizeof(InternalAlloc))) InternalAlloc(*state_proxy);
+    }
+
+    /**
+     * Predicate function for querying whether an object is entirely within the
+     * sandbox.
+     */
+    bool is_in_sandbox(void* ptr, size_t sz)
+    {
+      return (ptr >= start) && (pointer_offset(ptr, sz) < top);
+    }
+
+    /**
+     * Predicate function for querying whether an object is entirely within the
+     * region of the sandbox allocated for its heap.
+     */
+    bool is_in_sandbox_heap(void* ptr, size_t sz)
+    {
+      return (
+        ptr >= pointer_offset(start, sizeof(SharedState)) &&
+        (pointer_offset(ptr, sz) < top));
+    }
+
+  private:
+    void* alloc_sandbox_heap(size_t sb_size)
+    {
+      if constexpr (pal_supports<AlignedAllocation, DefaultPal>)
+      {
+        return DefaultPal::reserve_aligned<true>(sb_size);
+      }
+      else
+      {
+        return DefaultPal::reserve_at_least(sb_size).first;
+      }
+    }
+  };
+}
+
+int main()
+{
+  static const size_t sb_size = 128 * 1024 * 1024;
+
+  // Check that we can create two sandboxes
+  Sandbox sb1(sb_size);
+  Sandbox sb2(sb_size);
+
+  auto check = [](Sandbox& sb, auto& alloc, size_t sz) {
+    void* ptr = alloc.alloc(sz);
+    assert(sb.is_in_sandbox_heap(ptr, sz));
```

Comment:
SNMALLOC_CHECK rather than `assert`.  

FYI.  We use our own `assert` as it is not defined in OE.